src/scripting_engine.c

```diff
@@ -0,0 +1,284 @@
+ */
+int scriptingEngineManagerRegister(const char *engine_name,
+                                   ValkeyModule *engine_module,
+                                   engineCtx *engine_ctx,
+                                   engineMethods *engine_methods) {
+    sds engine_name_sds = sdsnew(engine_name);
+
+    if (dictFetchValue(engineMgr.engines, engine_name_sds)) {
+        serverLog(LL_WARNING, "Same engine was registered twice");
+        sdsfree(engine_name_sds);
+        return C_ERR;
+    }
+
+    scriptingEngineImpl *ei = zmalloc(sizeof(scriptingEngineImpl));
+    *ei = (scriptingEngineImpl){
+        .ctx = engine_ctx,
+        .methods = {
+            .create_functions_library = engine_methods->create_functions_library,
+            .call_function = engine_methods->call_function,
+            .free_function = engine_methods->free_function,
+            .get_function_memory_overhead = engine_methods->get_function_memory_overhead,
+            .get_memory_info = engine_methods->get_memory_info,
+        },
+    };
+
+    client *c = createClient(NULL);
+    c->flag.deny_blocking = 1;
+    c->flag.script = 1;
+    c->flag.fake = 1;
+
+    scriptingEngine *e = zmalloc(sizeof(*ei));
+    *e = (scriptingEngine){
+        .name = engine_name_sds,
+        .module = engine_module,
+        .impl = ei,
+        .c = c,
+        .module_ctx = engine_module ? moduleAllocateContext() : NULL,
+    };
+
+    dictAdd(engineMgr.engines, engine_name_sds, e);
+
+    engineMemoryInfo mem_info = scriptingEngineCallGetMemoryInfo(e);
+    engineMgr.engine_cache_memory += zmalloc_size(e) +
+                                     sdsAllocSize(e->name) +
+                                     zmalloc_size(ei) +
+                                     mem_info.engine_memory_overhead;
+
+    return C_OK;
+}
+
+/* Removes a scripting engine from the engine manager.
+ *
+ * - `engine_name`: name of the engine to remove
+ */
+int scriptingEngineManagerUnregister(const char *engine_name) {
+    dictEntry *entry = dictUnlink(engineMgr.engines, engine_name);
+    if (entry == NULL) {
+        serverLog(LL_WARNING, "There's no engine registered with name %s", engine_name);
+        return C_ERR;
+    }
+
+    scriptingEngine *e = dictGetVal(entry);
+
+    functionsRemoveLibFromEngine(e);
+
+    zfree(e->impl);
+    sdsfree(e->name);
+    freeClient(e->c);
+    if (e->module_ctx) {
+        serverAssert(e->module != NULL);
+        zfree(e->module_ctx);
+    }
+    zfree(e);
+
+    dictFreeUnlinkedEntry(engineMgr.engines, entry);
+
+    return C_OK;
+}
+
+/*
+ * Lookups the engine with `engine_name` in the engine manager and returns it if
+ * it exists. Otherwise returns `NULL`.
+ */
+scriptingEngine *scriptingEngineManagerFind(sds engine_name) {
+    dictEntry *entry = dictFind(engineMgr.engines, engine_name);
+    if (entry) {
+        return dictGetVal(entry);
+    }
+    return NULL;
+}
+
+sds scriptingEngineGetName(scriptingEngine *engine) {
```

Comment:
it would be ideal if we express the "constness" of the returned sds string explicitly. Similar callouts to other helper functions returning plain pointers. That said, since we have never enforced "constness" in the past in our code base I suspect this could turn into quite a code churn. So definitely not a blocker for this PR (and it is primarily for code hygiene).

```
typedef const char* sds_const; // in src/sds.h
```
```suggestion
sds_const scriptingEngineGetName(scriptingEngine *engine) {
```

Reply:
I like this idea! The const-sds type would be useful in some places where we now have to cast away the const modifier. Even if we don't introduce it here, we can do it in another PR. @SoftlyRaining I know you have run into this in another PR.

Reply:
I do like this! There have been various places I would like to have used it :)

Reply:
I agree that it would be nice have this constness property for sds strings. 
I suggest to open an issue for doing this improvement, because from a quick look at the code, it will require changes to code that is not related to this feature.

Reply:
separate PR dedicated to "constness" is better IMO.

Reply:
filed #1542.

Reply:
now that the `const_sds` pr is merged. let's make sure we make good use of it. separate pr is also fine.